tvx/state/surgeon.go

```diff
@@ -173,84 +175,80 @@ func (sg *Surgeon) pluckActorStates(tsk types.TipSetKey, pluck []address.Address
 }
 
 // saveInitActor saves the state of the init actor to the provided state map.
-func (sg *Surgeon) saveInitActor(initState *init_.State, stateMap *adt.Map) error {
+func (sg *Surgeon) saveInitActor(initActor *types.Actor, initState init_.State, stateMap *state.StateTree) error {
 	log.Printf("saving init actor into state tree")
 
 	// Store the state of the init actor.
 	cid, err := sg.stores.CBORStore.Put(sg.ctx, initState)
 	if err != nil {
 		return err
 	}
-	actor := &types.Actor{
-		Code: builtin.InitActorCodeID,
-		Head: cid,
-	}
+	actor := *initActor
+	actor.Head = cid
 
-	err = stateMap.Put(adt.AddrKey(builtin.InitActorAddr), actor)
+	err = stateMap.SetActor(init_.Address, &actor)
 	if err != nil {
 		return err
 	}
 
-	cid, _ = stateMap.Root()
+	cid, _ = stateMap.Flush(sg.ctx)
 	log.Printf("saved init actor into state tree; new root: %s", cid)
 	return nil
 }
 
 // retainInitEntries takes an old init actor state, and retains only the
 // entries in the retain set, returning a new init actor state.
-func (sg *Surgeon) retainInitEntries(oldState *init_.State, retain []address.Address) (*init_.State, error) {
+func (sg *Surgeon) retainInitEntries(s init_.State, retain []address.Address) (init_.State, error) {
 	log.Printf("retaining init actor entries for addresses: %v", retain)
 
-	oldAddrs, err := adt.AsMap(sg.stores.ADTStore, oldState.AddressMap)
-	if err != nil {
-		return nil, err
-	}
+	// TODO: The simplest approach would be to just add a "remove" method
+	// (as below) to the init actor abstraction. However, I'm not sure if we
+	// even _want_ this code. From what I can tell, lotus never removes
+	// address in the init actor, even if the underlying actor disappears.
```

Comment:
Big TODO. This is a weird case because:

1. We usually only _read_ state using these abstractions (we rarely modify).
2. We never delete actors from this map. So any function to do so would exist for oni (of course, maybe we _should_ be deleting actors?).

Reply:
Alternatively, we could introduce per-version logic.

Reply:
We only use this to trim down the state tree when we extract a test vector from an on-chain message. This constructs a synthetic state tree, with only the entries that the vector effectively needs. For a given message M, we don't need the _full_ address mapping inside the init actor.

Reply:
Ok, it turns out that `init_.State` is now an interface, and we can't freely manipulate the underlying struct. We're gonna have to type assert into a concrete struct.